src/psyclone/psyir/backend/fortran.py

```diff
@@ -0,0 +1,443 @@
+
+        :returns: The Fortran code as a string.
+        :rtype: str
+
+        '''
+        if not node.name:
+            raise VisitorError("Expected node name to have a value.")
+
+        result = (
+            "{0}module {1}\n"
+            "".format(self._nindent, node.name+"_mod"))
+
+        self._depth += 1
+        args = [symbol.name for symbol in node.symbol_table.argument_list]
+        result += (
+            "{0}use constants_mod, only : r_def, i_def\n"
+            "{0}implicit none\n"
+            "{0}contains\n"
+            "{0}subroutine {1}({2})\n"
+            "".format(self._nindent, node.name, ",".join(args)))
+
+        self._depth += 1
+        # Declare the kernel data.
+        declarations = ""
+        for symbol in node.symbol_table._symbols.values():
+            declarations += self.get_declaration(symbol)
+        # Get the executable statements.
+        exec_statements = ""
+        for child in node.children:
+            exec_statements += self.visit(child)
+        result += (
+            "{0}\n"
+            "{1}\n"
+            "".format(declarations, exec_statements))
+
+        self._depth -= 1
+        result += (
+            "{0}end subroutine {1}\n"
+            "".format(self._nindent, node.name))
+
+        self._depth -= 1
+        result += (
+            "{0}end module {1}\n"
+            "".format(self._nindent, node.name+"_mod"))
+        return result
+
+    def assignment(self, node):
+        '''This method is called when an Assignment instance is found in the
+        PSyIR tree.
+
+        :param node: An Assignment PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.Assigment`
+
+        :returns: The Fortran code as a string.
+        :rtype: str
+
+        '''
+        lhs = self.visit(node.children[0])
+        rhs = self.visit(node.children[1])
+        result = "{0}{1}={2}\n".format(self._nindent, lhs, rhs)
+        return result
+
+    def binaryoperation(self, node):
+        '''This method is called when a BinaryOperation instance is found in
+        the PSyIR tree.
+
+        :param node: A BinaryOperation PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.BinaryOperation`
+
+        :returns: The Fortran code as a string.
+        :rtype: str
+
+        '''
+        # reverse the fortran2psyir mapping to make a psyir2fortran
+        # mapping
+        from psyclone.psyGen import Fparser2ASTProcessor as f2psyir
+        mapping = {}
+        for operator in f2psyir.binary_operators:
+            mapping_key = f2psyir.binary_operators[operator]
+            mapping_value = operator
+            # Only choose the first mapping value when there is more
+            # than one.
+            if mapping_key not in mapping:
+                mapping[mapping_key] = mapping_value
+        lhs = self.visit(node.children[0])
+        oper = mapping[node._operator]
+        rhs = self.visit(node.children[1])
+        result = "{0}{1}{2}".format(lhs, oper, rhs)
```

Comment:
I think you still have an issue with the operators with a functional syntax (e.g.: BinaryOperation.Operator.REM -> Fortran `mod(a, b)`, this are not in `f2psyir.binary_operators`. Maybe you can have a look at how I implemented the BinaryOperator.gen_c_code()

Reply:
Well noticed. I've created a new issue #414 which should deal with mappings to/from Fortran. Your solution in `gen_c_code` was to create a new (reverse) mapping. I don't want to do this, so created the issue and came up with a stop gap solution.

Reply:
I am happy to offload this issue to #414 as I think the operation/intrinsic double direction map should be designed carefully to do everything we need.